### ISSUE TITLE
Add a export hint for NodePath and export Nodes

### ISSUE THREAD

**Issue by brndel** [godotengine/godot-proposals#934]
**Describe the project you are working on:**
Inventory system

**Describe the problem or limitation you are having in your project:**
I have an inventory control, that gets a reference to the inventory it shows using
`export(NodePath) var inventory_path`
`onready var inventory : Inventory = get_node_or_null(inventory_path)`

1. this is unnecessarily complicated
2. i can't be 100% sure, wether inventory_path refers to an Inventory
3. if anyone else wants to use my Inventory system, i would need to add get_configuration_warning for every export of NodePath, so he knows, that he should only select Inventory Nodes

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Export hints for NodePath and direct export of Nodes would be easier and safer

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
`export(NodePath, Spatial) var spatial_node`
this would only make Spatial nodes selectable in the 'Select a Node' dialog window

`export(NodePath, Inventory) var inventory`
this would only make nodes with the Inventory script selectable in the 'Select a Node' dialog window

`export(MeshInstance) var mesh_instance`
this would basically be the same as but simpler
`export(NodePath, MeshInstance) var mesh_instance_path`
`onready var mesh_instance : MeshInstance = get_node(mesh_instance_path)`

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Not really, exporting NodePaths is always a little bit risky

**Is there a reason why this should be core and not an add-on in the asset library?:**
This can't be achieved with add-ons

I dont know wether this is hardcoded, but when creating a ViewportTexture this dialog shows up, where you are only able to select Viewport Nodes.
![image](https://user-images.githubusercontent.com/58010307/83092121-f7259880-a09c-11ea-9c90-640bcc3405be.png)
I think it shouldn't be much work connecting this with the export type

**Comment by nhold** [comment]
I mean, just direct export would be best i.e:

`export meshInstance : MeshInstance`

**Comment by Mantissa-23** [comment]
You currently can't export reference types without an error anyways. I think @nhold's suggestion really makes the most sense and would be most intuitive for people.

**Comment by MartinHaeusler** [comment]
I'm coming from Unity, and I just hit this very same issue. I've got no clue about GDScript, but in Unity C# you can simply do this:

```
class MyPlayer: MonoBehaviour {

    public Camera playerCamera;

}
```

... which will bring up a `playerCamera` field in the editor of type **Camera** to which you can only assign cameras in the scene. There's no need to tell Unity that this needs to be a node path of sorts; it knows that already because the target type extends `GameObject`.

I see three ways Godot could do this.

# Way 1: Adding Generics to `NodePath`

The less-than-ideal way would be making NodePath generic, like so:

```
class MyPlayer: KinematicBody {

    [Export]
    public NodePath<Camera> playerCamera;

    private Camera _playerCamera;
 
    public override _Ready() {
        this._playerCamera = this.getNode(this.playerCamera);
    }
}
```

Pro:
 - relatively simple
 - type safe

Con:
 - Need to do `this.getNode(...)`
 - Need to have a second field
 - Simply unwieldy

# Way 2: Using export hints

```
class MyPlayer: KinematicBody {

    // note the hint. This tells Godot editor that the inspector should contain a path.
    [Export(hint=PropertyHint.NODE_PATH)]
    public Camera playerCamera;

}
```

Pro:
 - still quite simple
 - type safe
 - no need for additional fields or manual `getNode(...)` calls.
 - Godot editor can easily check if the field should actually contain a node path or not.

Con:
 - This will trip up newcomers who don't know about the hint.
 - It's duplicate information, because we will ALWAYS want the hint on all fields which have a type that inherits from `Node`.

# Way 3: Inferring the hint from the field type

This is the way Unity handles it, and it's the best way IMHO because it requires zero effort or knowledge from the user:

```
class MyPlayer: KinematicBody {

    [Export]
    public Camera playerCamera;

}
```

Godot already knows that `Camera` inherits from `Node`. So the editor knows that a node path is required. And it also knows what kind of node is expected due to the field type. The information is all there; the editor and runtime simply need to be smart enough to pick it up.

**I would strongly vote for way 3.** Unity is worth copying in this case because this is one of the things that work extraordinarily well.

**Comment by Jummit** [comment]
3 seems like the cleanest way.
`export player_camera : Camera`

**Comment by neikeq** [comment]
>```
>[Export]
>public NodePath<Camera> playerCamera;
>```

Making NodePath generic doesn't sound like a good idea to me and it has little use other than this example. Probably something like this would be better:

```cs
[Export(typeof(Camera))]
public NodePath playerCamera;
```

>```
>[Export(hint=PropertyHint.NODE_PATH)]
>public Camera playerCamera;
>```

This one would probably be better as something like OnReadyExport:

```cs
[OnReadyExport]
public Camera playerCamera;
```

More complicated to implement but possible. It's something I've wanted to add for some time.

**Comment by nhold** [comment]
In what use case would you use `[Export]` instead of `[OnReadyExport]`?

**Comment by me2beats** [comment]
This is my attempt to partially solve it with a plugin.
https://github.com/me2beats/nodepath-filter

Implemented very hacky (ugly), but seems to work (except some known issues) and can be used as a temporary workaround.

**Comment by KoBeWi** [comment]
Exporting node type is implemented in 4.0
